### ISSUE TITLE
[Question] How to bypass hard ssh requirement for Multi-node distributed training

### ISSUE THREAD

**Issue by Randolph-zeng** [microsoft/DeepSpeed#1603]
Hi DeepSpeed team:

I was trying to use deepspeed in a **slurm environment where ssh connections between the allocated computing resources is not 
possible.** It seems like, as one of your great teammate @jeffra mentioned [in this issue ](https://github.com/microsoft/DeepSpeed/issues/274#issuecomment-648611279), it is a hard requirement to have password-less ssh connections between nodes. 

Is there any tutorial available that I can only expose some ports in the nodes to use deepspeed for distributed training? I feel like this should be a common scenario but I could not find any solution to this in the internet. 

I will appreciate any pointer if possible. Thanks a lot!

**Comment by jeffra** [comment]
Hi @Randolph-zeng, my previous comments on this hard requirement should be clarified :) If you want to use our deepspeed launcher then you'll need password-less ssh to work. _However_, if you are able to launch processes with MPI then you can still use DeepSpeed (just without our launcher). Please see these details on DeepSpeed support with MPI launching: https://www.deepspeed.ai/getting-started/#mpi-and-azureml-compatibility

**Comment by Randolph-zeng** [comment]
Hi @jeffra , thanks for the speedy response. After numerous trial and error, I got the inspiration from a torch lightning example where indeed a deepspeed launcher is not used. Instead a plain `torch.distributed.init_process_group`(with init method being env) is used before `deepspeed.initialize`. I am not sure why this works, but I will close this issue now. Hopes this can help other  users who might have same problems.
PS: For my curiousity would you be kind enough to shed some insight for me(or perhaps future users) why this work ? Such as 
1. why use torch init_process_group instead of the recommeneded ds.init_distributed would work?
2. how does deepspeed communicate in this fashion, by a port instead of ssh? Why the launcher requires ssh, is it because of the nccl backend ?
3. how did the master node know each other without explicitly passing in the ip addresses of other nodes , isn't deepspeed using round robin gradient reduce under the hood ? Or is it a single direction communication where only the master ip address is required ? 
Thanks a lot for your time

**Comment by Randolph-zeng** [comment]
Turns out the method I mentioned above did not quite solve the issue, it allows communication between two nodes, but only one GPU in each node is invoked. Am still debugging and exploring ....

**Comment by jeffra** [comment]
In our case the deepspeed launcher pre-dates the `torchrun`/elastic-launcher in newer pytorch. I don't believe we've tried the new launcher to know for sure here.

Does `torchrun` launch processes across multiple nodes when invoked from a single machine? Often we see (such as with AzureML) that launchers like `mpirun` will be invoked from the infrastructure side for the user and then it is responsible for launching the initial processes on each machine. In our case we often have _interactive_ SSH access to several nodes and can launch many new multi-node training jobs in order to reuse the resources we're allocated. So in this case we simply use this SSH ability to do the initial per-node process launching and then each node will create N sub-processes that are bound to each GPU on each node (e.g., N=4 if a single machine has 4 GPUs on it). 

I should stress that the SSH/PDSH usage is only used at the start to launch the training processes and then from that point forward we rely on torch + NCCL for all communication collectives (e.g., all-reduce, all-gather). Nothing performance critical is happening over SSH.

I hope this answers your questions, feel free to clarify if I am misunderstanding.

**Comment by Randolph-zeng** [comment]
Hi Jeff, Really appreciate the insight you provide! That really helps to clarified the problems. To further clarify the approach I used, I am not exploring the openmpi route any more. Instead, I am within a SLURM environment, where I am creating one task per node, and relying on this task to initialize all the workers within this node. This approach is pretty similar to what you have described that DS launcher will do via ssh. It's just the SLURM doing the launch job for me, since it forbit ssh across nodes for safety purpose.
For future comers who might have similar problem, I will post my solution here, hope that makes your life a bit easier : )
This is the slurm sbatch script head I am using for a 2-node, 2 gpus per node, 4 workers in total toy example
```
#!/bin/sh -l
#SBATCH -J txltest
#SBATCH -p batch
#SBATCH --ntasks-per-node=1
#SBATCH --nodes=2
#SBATCH --gres=gpu:2
#SBATCH --mem-per-cpu=8G # memory per cpu-core (4G is default)
#SBATCH --cpus-per-task=2
```
Below is the launcher I use to start my python script. Note that the `nnodes` should equal to the nodes of your slurm head. However, the `nproc_per_node` should equal to your gpu numbers per node for correct world_size calculation(torch run will calculate these for you and store them in your os.env). Also, node_rank is another very import param to pass in, in my case it will be 0 or 1.  Don't forget that MASTER_ADDR should be unique across all nodes : ).
```
export LAUNCHER="python -u -m torch.distributed.run \
        --nnodes=2 \
        --nproc_per_node=2 \
        --node_rank=$node_rank \
        --rdzv_id=$SLURM_JOBID
        --rdzv_backend=c10d
        --rdzv_endpoint=$MASTER_ADDR:6000
"
```
I will close this issue for now. Hope this helps.

**Comment by deema-A** [comment]
> MASTER_ADDR

    ```
raise RendezvousTimeoutError
torch.distributed.elastic.rendezvous.api.RendezvousTimeoutError
``` any advice?
